keps/draft-test-composability.md

```diff
@@ -0,0 +1,98 @@
+---
+kep-number: 0
+short-desc: This KEP proposes a scheme for making test steps, asserts, and manifests composable in order to prevent repetition of YAML across test cases.
+title: KUTTL Test Composability
+authors:
+  - "@jbarrick-mesosphere"
+owners:
+  - "@jbarrick-mesosphere"
+editor: TBD
+creation-date: 2020-02-12
+last-updated: 2020-02-12
+status: draft
+see-also:
+  - KEP-2
+---
+
+# KUTTL Test Composability
+
+This KEP proposes a scheme for making test steps, asserts, and manifests composable in order to prevent repetition of YAML across test cases.
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+* [Proposal](#proposal)
+
+## Summary
+
+KUTTL tests can tend to be quite repetitive, for example, for test case environment setup or if there are any common resources across test cases. Additionally, sometimes it is useful to compose common asserts.
+
+## Motivation
+
+This is KEP intends to improve the developer experience of writing tests, make tests easier to maintain and reduce duplication by making the files within them more composable. For example, one user [indicated that the tests are overly verbose](https://github.com/kudobuilder/kudo/issues/1311#issuecomment-580709826). Additionally, because this duplication can be expensive, some users deploy common resources into a single namespace breaking namespace isolation and it can make orchestrating tests difficult.
+
+Beyond just composability, the magic file name scheme indicating step index and whether or not a file is an `assert` or `errors` file can be opaque and hard to understand for new users. The proposal in this KEP will also make test structure easier to understand.
+
+### Goals
+
+* Improve composability of test steps, manifests, and asserts.
+* Improve reusability of YAML files.
+* Provide better integration with existing repository structures (e.g., deploy operator YAMLs from a release directory).
+* Make test file structure easier to understand.
+
+## Proposal
+
+The `TestStep` resource will be extended to allow users to provide manifest, errors, and assert files as a list. Three settings will be added: `apply` , `assert`, `error`. Each is a list of files that can either be a single file name or a directory to recurse over.
+
+* `apply`: a list of files to apply for the test step.
+* `assert`: a list of files containing test asserts to expect.
+* `error`: a list of files containing test errors that will cause the test to fail if their state is observed.
```

Comment:
Does it make sense to also have a `delete` resources, to complement the `apply` step?

Reply:
Oh, I see there is already a `delete` section...

Reply:
Yeah, but that would be a good thing to mention here.